### DISCUSSION TITLE
Transport / transit network integration

### DISCUSSION THREAD

**Discussion by overflorian** [organicmaps/organicmaps#6275] [Ideas]
Hi, following a discussion with @biodranik, let's explore ways to integrate transit networks in the app.

# Goals
We should aim to produce a better transport map render and routing experience than proprietary software, using crowdsourcing and open data as leverage.

To achieve that, we need to:
1. use external data from the transit authorities and companies, known as GTFS files, to provide static geographic and schedule information
2. use and improve OSM data in order to generate a better geographical data quality than the one existing in each GTFS
3. use external APIs from transit authorities and companies to provide real-time data
4. improve the map render to display transit data and improve the routing engine

# Steps
I suggest progressive goals to achieve this long-term vision:
**1. use existing OSM data**
 - clicking on a bus stop should display the number, name of the line, and the direction (destination)
_Example: clicking on this bus stop https://www.openstreetmap.org/node/6263448326 should display (line number 513, line name "La Traverse Bièvre Montsouris" and destination "Alésia - Général Leclerc" that is displayed in the relation https://www.openstreetmap.org/relation/1086658_
- activating a new "bus layer" should display the shape and color of each bus route and stops
Later in the process, add other transports than a bus, including but not limited to train, tram, subway...
This is approximately how it's done in OSMand, and it should be replicated but with a simpler UX in Organic Maps.

**2. use external GTFS files**
These "officially produced" files apply to a specific city / region and should be specifically handled.
This implies a massive amount of sub-constraints, such as and not limited to 
 - being able to link the GTFS data to OSM data with a perenne and unique ID
_For example this bus stop https://www.openstreetmap.org/node/6263448326 has a unique ID ref:FR:STIF=28223 that allows to link it in a unique way to the local authority referential ([here](https://data.iledefrance-mobilites.fr/explore/dataset/arrets-transporteur/table/?disjunctive.fournisseurname&q=28223&dataChart=eyJxdWVyaWVzIjpbeyJjb25maWciOnsiZGF0YXNldCI6ImFycmV0cy10cmFuc3BvcnRldXIiLCJvcHRpb25zIjp7ImRpc2p1bmN0aXZlLmZvdXJuaXNzZXVybmFtZSI6dHJ1ZSwicSI6IjI4MjIzIn19LCJjaGFydHMiOlt7ImFsaWduTW9udGgiOnRydWUsInR5cGUiOiJsaW5lIiwiZnVuYyI6IkFWRyIsInlBeGlzIjoiYXJ0eGVwc2cyMTU0Iiwic2NpZW50aWZpY0Rpc3BsYXkiOnRydWUsImNvbG9yIjoiIzY0QjVFNSJ9XSwieEF4aXMiOiJhcnRjaGFuZ2VkIiwibWF4cG9pbnRzIjoiIiwidGltZXNjYWxlIjoieWVhciIsInNvcnQiOiIifV0sImRpc3BsYXlMZWdlbmQiOnRydWUsImFsaWduTW9udGgiOnRydWV9))_
 - quality validation
 - prioritization of the data used between OSM and GTFS, according to specific rules
 - ...

**3. improve GTFS quality in OSM**
This would allow any user to improve OSM data on the top of the GTFS file, in order for OSM to always be a better data source.

**4. use real-time data**
The last step would be to use real-time data from external APIs, with the need for an internet connection for the user.

---
My company [Jungle Bus](https://github.com/Jungle-Bus) contributed to produce a lot of open-source software to solve these issues and we are happy to help.
Please see https://wiki.openstreetmap.org/wiki/Public_transport/Tools for more information. There is a page dedicated to quality control: https://wiki.openstreetmap.org/wiki/Public_transport/Quality_Assurance

PS: in order to achieve the vision of OSM being inevitable for bus transport over the large-Paris-area, we 
1. mapped every single bus stop in OSM, including a unique ID (it took more than 4 years to achieve this first step)
2. contacted the local authority, which was happy to request a detailed audit to compare OSM VS GTFS data ([here, only in French](https://junglebus.io/les-donnees-de-transport-en-ile-de-france-dans-openstreetmap-un-potentiel-largement-inexploite)), that took one more step
3. chose the "OSM first" approach for our future projects ([see the full infographic here](https://junglebus.io/en/osm-then-gtfs-or-gtfs-then-osm/)) to make sure OSM would remain ahead of other data sources.
4. work with the local authority to improve the GTFS quality using OSM ([see open data comparing the distance of the bus stops here](https://data.iledefrance-mobilites.fr/explore/dataset/ecarts-arrets-referentiel-et-openstreetmap/custom/))

We would be happy to be an experimental territory to test any step for Organic Maps to move forward.

**Comment by patepelo** [comment]
I'm happy to read this. Public transport is something very important that need improvement on apps based on OSM. 

This project is worth mentioning: https://github.com/grote/Transportr

I believe they are newly active on development again.

**Comment by biodranik** [comment]
See related issues: https://github.com/organicmaps/organicmaps/issues?q=is%3Aissue+is%3Aopen+gtfs+label%3A%22Public+Transport%22

A preliminary vision for the OM's PT implementation:
1. Validate and generate bus, tram, rail, and ferry lines information from OSM, in a similar way as it is done now for [subways](https://cdn.organicmaps.app/subway/) (@alexey-zakharenkov may help with his [validator](https://github.com/alexey-zakharenkov/subways)) or using some other approach that allows to display and build simple routes in OM without schedule/real-time information. It should cover many cities and motivate the OSM community to update the transport network.
2. Collect real-time GTFS feeds on a server, map them to OSM data (the most complex task btw), and efficiently distribute them to clients (likely in a simple custom format suitable for the routing algorithm) so they can cache this data and use it even in offline when building routes.

Any volunteers to help/lead the implementation are welcome. The goal is to build a public transport service that is better than existing ones.

**Comment by jumelles** [comment]
GTFS definitely seems like the way to go. The [GTFS site](https://gtfs.org/) explains how it covers more than just timetables:

> GTFS is most widely known for trip planning information, particularly in metro areas with fixed-route service. However, there are a variety of optional features above and beyond the basic GTFS Schedule specification that make GTFS more widely applicable, including Fares for showing fare costs and structures, Flex (in development) for demand-responsive transit options, such as dial-a-ride and paratransit services, and Pathways for displaying accessibility information that’s vital to riders using mobility devices or needing additional accommodations. GTFS Realtime builds upon GTFS Schedule and on-vehicle GPS systems to provide real-time updates on vehicle locations.

If I'm understanding correctly, it would make sense to pre-download the GTFS Schedule, and then use GTFS Realtime when Organic Maps has an internet connection.

**Comment by overflorian** [comment]
Hi, I agree with step 1. suggested by @biodranik.

Step 2. is more complicated... and should be discussed.

2 scenarios:
1. mine: use GTFS as they are, as you suggest @jumelles 
2. yours @biodranik: Transform GTFS data to OSM format

Each option has its benefits and drawbacks:
1. GTFS: by-default management of timetables and more (mobile booking and payment...), need a proper server-side repository with OpenTripPlanner or Valhalla
2. OSM format: already supported in OM, less short-term development but need specific software to be supported such as GTFS2OSM. Also it might be limiting in the future.

**Reply by biodranik** [reply]
It may make sense to stop limiting the implementation to only existing solutions and not be afraid to implement something if necessary. We would never create MapsWithMe/Maps.Me/Organic Maps if we stick to existing solutions.

**Reply by TheMrBooblik** [reply]
Hi @biodranik, it would be amazing to start with less complex solution for the first time - take `Operator` and `Network` fields from OSM and show them when you clicking on bus/tram/etc stop

<img width="620" alt="Screenshot 2023-10-26 at 15 54 17" src="https://github.com/organicmaps/organicmaps/assets/21023603/ce51039d-7379-4031-b2ea-28e8768bd419">

I could give an example of importance of this update in Italy, sometimes here you can't understand which stop is belong to which bus operator. And google maps don't have most bus stops in Calabria.

**Reply by biodranik** [reply]
That is an easier step, yes. Adding and displaying these fields and route numbers is a good basic step. Any volunteers?

**Reply by patepelo** [reply]
OsmAnd is already beta testing with these feature.

By the way, we should already consider separating the stop position (bus=yes + public_transport=stop_position) from the bus platform (bus=yes + highway=bus_stop + public_transport=platform)

Currently it is like these:
![image](https://github.com/organicmaps/organicmaps/assets/23003127/6be4e422-db8c-4c9a-b526-4c43ddcffc5f)


My idea of the first implementation would be to be able to click the bus stop and see the lines that go past it. If there are two exact same POIs it would be difficult to know which one to click. The one with the platform is afaik not tagged to the relation, so should be considered similar to a subway entrance or even less, since it wouldnt have much relevance since it is always so close to the stop position.

**Reply by biodranik** [reply]
Let's create a separate issue for stop position, with the best examples form other apps.

**Reply by PinguDEV-original** [reply]
@biodranik where can I find the issue?

**Reply by biodranik** [reply]
Is this one related? https://github.com/organicmaps/organicmaps/issues/2175

Displaying existing OSM metadata for existing OSM stops would already be very helpful.

**Comment by amirshahab89** [comment]
Hi! This is an awesome idea & I hope to see it come to fruition soon in the near future.

Here in Sweden, there is "Trafiklab". It seems that they have the full coverage of all public transport in the nation with relevant companies. See here; https://www.trafiklab.se/ 

They mention these GTFS Files can be used here; https://www.trafiklab.se/docs/using-trafiklab-data/using-gtfs-files/ 

The "GTFS Regional" API is live, as they mention here; https://www.trafiklab.se/api/ 

Hope someone here in the community knows how to use/ implement these into OM the best suitable way possible.

**Comment by patepelo** [comment]
Just adding some references that might help in the future:

https://overpass-api.de/ --> https://overpass-api.de/public_transport.html

Example: --> https://overpass-api.de/api/sketch-line?network=VRR&ref=625&operator=


And also consider the efforts of the[ Trufi Assosiacation](https://www.trufi-association.org/), also on [GitHub](https://github.com/trufi-association/)

**Comment by horvbalint** [comment]
I am not sure, but wouldn't this new project (started few weeks ago at fossdem) when ready, help organic maps to integrate public transport?
Maybe some collaboration would be beneficial :)
https://github.com/public-transport/transitous

**Comment by Machion** [comment]
I would be very happy, if at least line numbers soon find it's way into OM. It would be very helpful

**Comment by eagle-programming** [comment]
Directions for the Underground and other new public transport should be displayed. Otherwise you won't know which line to take. Data should be available from OSM.
![Screenshot_20240919-111021](https://github.com/user-attachments/assets/3dd486fe-0efa-4f67-ac1f-ea21dd93e171)
I only know to take the U2 and U3, but not the direction.

**Comment by itskme** [comment]
There is an offline, FOSS implementation that is remarkably global and accurate that already exists, Transito. 

I encourage the team to have a good look at this: 

https://git.sr.ht/~mil/transito

It even works in my small Canadian city with their transit agency. Offline!
